game/src/Ai/Behaviors/Tank/TankBTs.cpp

```diff
@@ -0,0 +1,843 @@
+	{
+		if(tankComp.rotateLeft)
+		{
+			tankComp.rotateLeft = false;
+		}
+		else
+		{
+			tankComp.rotateLeft = true;
+		}
+	}
+	//Rotate towards target end
+}
+void TankBT::rotateTowards(Entity entityID, glm::vec3 target, float rotSpeed, float precision)
+{
+	TankComponent& tankComp = getTankComponent();
+	Transform& tankTrans = getTankTrans();
+	//Rotate towards target start
+	tankTrans.updateMatrix();
+	glm::vec2 targetPos			= glm::vec2(target.x, target.z);
+	glm::vec2 tankPos			= glm::vec2(tankTrans.position.x, tankTrans.position.z);
+	glm::vec2 curRot			= -glm::normalize(glm::vec2(tankTrans.forward().x, tankTrans.forward().z));
+	glm::vec2 tank_to_friend	= glm::normalize(targetPos - tankPos);
+
+	float angle_between			= glm::degrees(glm::acos(glm::dot(tank_to_friend, curRot)));
+	tankComp.tempRotAngle = angle_between;
+
+	if(tankComp.rotateLeft && angle_between >= precision)
+	{
+		tankTrans.rotation.y += rotSpeed * get_dt();
+	}
+	else if(angle_between >= precision)
+	{
+		tankTrans.rotation.y -= rotSpeed * get_dt();
+	}
+
+	//Check if we rotated in correct direction
+	tankTrans.updateMatrix();
+	targetPos			= glm::vec2(target.x, target.z);
+	tankPos				= glm::vec2(tankTrans.position.x, tankTrans.position.z);
+	curRot				= -glm::normalize(glm::vec2(tankTrans.forward().x, tankTrans.forward().z));
+	tank_to_friend		= glm::normalize(targetPos - tankPos);
+	angle_between		= glm::degrees(glm::acos(glm::dot(tank_to_friend, curRot)));
+	//If angle got bigger, then change direction
+	if(tankComp.tempRotAngle < angle_between)
+	{
+		if(tankComp.rotateLeft)
+		{
+			tankComp.rotateLeft = false;
+		}
+		else
+		{
+			tankComp.rotateLeft = true;
+		}
+	}
+	//Rotate towards target end
+}
+
+bool TankBT::rotationDone(Entity entityID, glm::vec3 target, float rotSpeed, float precision)
+{
+	Transform& tankTrans = getTankTrans();
+	tankTrans.updateMatrix();
+	glm::vec2 targetPos			= glm::vec2(target.x, target.z);
+	glm::vec2 tankPos			= glm::vec2(tankTrans.position.x, tankTrans.position.z);
+	glm::vec2 curRot			= -glm::normalize(glm::vec2(tankTrans.forward().x, tankTrans.forward().z));
+	glm::vec2 tank_to_friend	= glm::normalize(targetPos - tankPos);
+
+	float angle_between			= glm::degrees(glm::acos(glm::dot(tank_to_friend, curRot)));
+
+	if(angle_between <= precision)
+	{
+		return true;
+	}
+	return false;
+}
+
+void TankBT::registerEntityComponents(Entity entityId)
+{
+	addRequiredComponent<TankComponent>(entityId);
+	addRequiredComponent<AiCombatTank>(entityId);
+}
+
+
+BTStatus TankBT::HasFreindsTarget(Entity entityID)
+{
+	BTStatus ret = BTStatus::Failure;
+	TankComponent& tankComp = getTankComponent();
+	if(tankComp.firendTarget.id != -1)
+	{
+		ret = BTStatus::Success;
+	}
+	return ret;
+}
+
+BTStatus TankBT::AreFriendsAlive(Entity entityID)
+{
+	BTStatus ret = BTStatus::Failure;
+	TankComponent& tankComp = getTankComponent();
+	if( tankComp.allFriends.size() > 0)
+	{
+		ret = BTStatus::Success;
+	}
+	/*for(auto f: tankComp.allFriends)
```

Comment:
mycket ut kommenterad kod, men jag kan se varför du kanske inte kan lägga in det nu